src/app/services/voting.service.js

```diff
@@ -149,7 +154,15 @@ class Voting {
         const signTransaction = (i) => {
             let p;
             if (common.isHW) {
-                p = account.signTransaction(broadcastData.transactions[i]).first().toPromise();
+                if (this._Wallet.algo == "trezor") {
+                    p = account.signTransaction(broadcastData.transactions[i]).first().toPromise();
+                } else if (this._Wallet.algo == "ledger") {
+                    const transaction = broadcastData.transactions[i];
+                    transaction.signer = nem.PublicAccount.createWithPublicKey("462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce");
```

Comment:
Could you elaborate where is this publicKey comes from? `462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce`

Reply:
Currently, the NEM-sdk hasn't support for hardware while preparing the transaction object yet when running the nem.model.transaction.prepare, for example when we create mosaic this function has no data of mosaic creator (ref [here](https://github.com/QuantumMechanics/NEM-sdk/blob/a9d93b1bced51c0c2c6838a01e233896da1dd847/src/model/transactions/mosaicDefinitionTransaction.js#L19-L21)) cause the wallet doesn't have private key of ledger device. So we have to correct the data of signer like this. Maybe some modification in NEM-sdk will fix it

Reply:
Is it needed for Trezor though ? I don't get why it is different for it under Electron environment. `Trezor.serialize` does not correct the signer like `Ledger.serialize`.

Reply:
1. The voting NEM-service use [nem-voting](https://github.com/shierve/nem-voting/tree/master/src) to sign voting transactions. It uses a pop-up method to sign many transactions in 1 session which not work for Electron app (it keep the session open). Therefore we re-use the normal pop-up (sign only one transaction each time) to address this pop-up method for Electron build.
2. On Trezor, the signer is correct on [trezor-firmwalre](https://github.com/trezor/trezor-firmware/blob/7d8fbacba7c5945ca93068f6c44e0df134e6b657/crypto/nem.c#L367-L369). For Ledger use do that work on wallet before send it to Ledger devices.

Reply:
Right got it, however the code in Trezor firmware correct the public key if `!signer` but `462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce` is used here, shouldn't it be empty then ?

Reply:
We just update signer for mosaic creation transaction (ref [here](https://github.com/FDS-Ledger/NanoWallet/blob/d3734feceaa37fbc220317eb04d995fd29a62134/src/app/modules/ledger/ledger.service.js#L236-L242)) so no signer correction for bot Trezor and Ledger wallets now. Do you mean that there's something wrong with voting signing and creating flow on hardware wallets?

Reply:
> Do you mean that there's something wrong with voting signing and creating flow on hardware wallets?

Well it seems something is wrong with voting, a public key is set (`462ee976890916e54fa825d26bdd0235f5eb5b6a143c199ab0ae5ee9328e08ce`) but is not corrected maybe it is somewhere but I can't find it anywhere. I am completely lost between several files and libraries.